### ISSUE TITLE
Data source based on OSS backend doesn't support nested parameters  .assume_role["role_arn"]: an object is required.

### ISSUE THREAD

**Issue by hayorov** [hashicorp/terraform#29305]
### Terraform Version
```
Terraform v1.0.4
on linux_amd64
```

### Terraform Configuration Files

```terraform
data "terraform_remote_state" "level1" {
    backend   = "oss"
    config    = {
      bucket                = "XXX-XXX-XX"
      prefix                = "level1/networking"
      key                   = "terraform.tfstate"
      acl                   = "private"
      encrypt               = true
      region                = "ap-southeast-1"
      access_key = jsondecode(chomp(file(".provider_access"))).access_key
      secret_key = jsondecode(chomp(file(".provider_access"))).secret_key
      assume_role  = {
        role_arn = "acs:ram::XXX:role/resourcedirectoryaccountaccessrole"
      }
    }
}
```

### Debug Output
```
The given configuration is not valid for backend "oss":
.assume_role["role_arn"]: an object is required.
```

### Expected Behavior
Data source works as described in documentation with parameters defined in nested configuration block.

### Actual Behavior
A Validation error raises on any parameter used nested `assume_role` block. For example `.assume_role["role_arn"]: an object is required.`

### Steps to Reproduce
1. create file remote.tf with context 
  ```terraform
    data "terraform_remote_state" "level1" {
        backend   = "oss"
        config    = {
          bucket                = "XXX-XXX-XX"
          prefix                = "level1/networking"
          key                   = "terraform.tfstate"
          acl                   = "private"
          encrypt               = true
          region                = "ap-southeast-1"
          access_key = "XXX"
          secret_key = "YYY"
          assume_role  = {
            role_arn = "acs:ram::XXX:role/resourcedirectoryaccountaccessrole"
          }
        }
    }
  ```
2. run `terraform init`

### Additional Context

Official documentation https://www.terraform.io/docs/language/settings/backends/oss.html#data-source-configuration

Remote backend (OSS) works fine. 
Meanwhile, `data source` has its own schema mapping/validation, and based on my research the problem happens in the `data_source_state` during config to schema mapping.

```
# internal/builtin/providers/terraform/data_source_state.go
	configVal, err := schema.CoerceValue(config)  // <- here 
	if err != nil {
		diags = diags.Append(tfdiags.AttributeValue(
			tfdiags.Error,
			"Invalid backend configuration",
			fmt.Sprintf("The given configuration is not valid for backend %q: %s.", backendType,
				tfdiags.FormatError(err)),
			cty.Path(nil).GetAttr("config"),
		))
		return nil, cty.NilVal, diags
	}

```

```
# internal/configs/configschema/coerce_value.go
func (b *Block) coerceValue(in cty.Value, path cty.Path) (cty.Value, error) {
	switch {
	case in.IsNull():
		return cty.NullVal(b.ImpliedType()), nil
	case !in.IsKnown():
		return cty.UnknownVal(b.ImpliedType()), nil
	}

	ty := in.Type()
	if !ty.IsObjectType() {
		return cty.UnknownVal(b.ImpliedType()), path.NewErrorf("an object is required"). // <- here
	}

```

Also, there is no other backend that declares nested block under configuration. A good example is the `s3 backend`, which has similar functionality with assume_role and has a flat parameters design. Ref https://www.terraform.io/docs/language/settings/backends/s3.html#assume-role-configuration

```
assume_role_duration_seconds, assume_role_policy, assume_role_transitive_tag_keys
```

So it seems that this feature has never worked since the initial implementation. Based on other backend designs, the validation/schema mapping doesn't support nested blocks under the configuration section of the backrnd.

### References
#23755 closed as "This issue doesn't seem to be following one of the standard-issue templates..."

**Comment by hayorov** [comment]
UPD: if the design is acceptable I'm happy to add OSS backend tests and cover assume_role logic, currently there are no tests.

**Comment by hayorov** [comment]
Thx, @jbardin. @xiaozhu36 could you please take a look?

**Comment by hayorov** [comment]
Hello, any news here?

**Comment by xiaozhu36** [comment]
HI @hayorov There is an incompatible update and please have a check.

**Comment by hayorov** [comment]
> HI @hayorov There is an incompatible update and please have a check.

Hello, it's now fully backward compatible and I've resolved all your comments, please take a look.

**Comment by hayorov** [comment]
Many thanks @apparentlymart it's a good explanation and tbh I guessed during the implementation of https://github.com/hashicorp/terraform/pull/29307.

It seems that the same problem/question was asked before in https://github.com/hashicorp/terraform/issues/23755 and structures `[{}, ]` doesn't look organic (IMO).

I've supported a flattened variant with backward compatibility for the existing format and got maintainer approval here 
https://github.com/hashicorp/terraform/issues/23755

I'm sure that generalization across backends is a positive thing and will make the OSS backend more homogenous.
I Hope, you (the Hashi team) could assist with merging and I'm happy to close this issue. 

Once again, many thanks @apparentlymart @jbardin and @xiaozhu36.
